tree-sitter-stack-graphs/src/lib.rs

```diff
@@ -432,49 +434,130 @@ impl StackGraphLanguage {
-        parser.set_language(self.language)?;
+        parser.set_language(self.sgl.language)?;
         unsafe { parser.set_cancellation_flag(cancellation_flag.flag()) };
-        let tree = parser.parse(source, None).ok_or(LoadError::ParseError)?;
-
+        let tree = parser
+            .parse(self.source, None)
+            .ok_or(LoadError::ParseError)?;
         let parse_errors = ParseError::into_all(tree);
         if parse_errors.errors().len() > 0 {
             return Err(LoadError::ParseErrors(parse_errors));
         }
         let tree = parse_errors.into_tree();
 
-        let mut graph = Graph::new();
+        let mut globals = Variables::nested(globals);
+        let root_node = self.inject_node(NodeID::root());
         globals
-            .add(ROOT_NODE_VAR.into(), graph.add_graph_node().into())
+            .add(ROOT_NODE_VAR.into(), root_node.into())
             .expect("Failed to set ROOT_NODE");
+        let jump_to_scope_node = self.inject_node(NodeID::jump_to());
         globals
-            .add(JUMP_TO_SCOPE_NODE_VAR.into(), graph.add_graph_node().into())
+            .add(JUMP_TO_SCOPE_NODE_VAR.into(), jump_to_scope_node.into())
             .expect("Failed to set JUMP_TO_SCOPE_NODE");
+        let file_name = format!("{}", &self.stack_graph[self.file]);
         globals
-            .add(
-                FILE_PATH_VAR.into(),
-                format!("{}", &stack_graph[file]).into(),
-            )
+            .add(FILE_PATH_VAR.into(), file_name.into())
             .expect("Failed to set FILE_PATH");
-        let mut config = ExecutionConfig::new(&self.functions, &globals)
+
+        let mut config = ExecutionConfig::new(&self.sgl.functions, &globals)
             .lazy(true)
             .debug_attributes(
                 [DEBUG_ATTR_PREFIX, "tsg_location"].concat().as_str().into(),
                 [DEBUG_ATTR_PREFIX, "tsg_variable"].concat().as_str().into(),
             );
-        self.tsg
-            .execute_into(&mut graph, &tree, source, &mut config, &cancellation_flag)?;
 
-        let mut loader = StackGraphLoader::new(stack_graph, file, &graph, source);
-        loader.load(cancellation_flag)
+        // The execute_into() method requires that the reference to the tree matches the lifetime
+        // parameter 'a of the Graph, because the Graph can hold references to the Tree. In this Builder,
+        // the Graph is created _before_ the Tree, so that it can be prepopulated with nodes. Because of
+        // that, the borrow checker complains that the Tree only lives as long as this method, not as long
+        // as the lifetime parameter 'a. Here we transmute the Tree reference to give it the required 'a
+        // lifetime, which is safe because:
+        // (1) this method takes ownership of the Builder; and
+        // (2) it returns no values connected to 'a.
+        // These together guarantee that no values connected to the lifetime 'a outlive the Tree.
```

Comment:
I tried that, even though I wasn't a big fan of having the work split over creating the builder and executing it. But, it didn't work. Because we own the `tree`, any reference we take will have a smaller lifetime than the `'a` parameter.

I agree that it would be nicer to be able to avoid this, but I haven't figured out how that is possible.

Reply:
Got it.  I think there have been discussions (and maybe even PRs) about cleaning up the lifetimes in tree-sitter proper.  That might be it easier to untangle these kinds of things.  But that's in the future, :+1: for this approach in this PR for now!